### PR TITLE
Add support for `sockaddr_un` on Windows.

### PR DESCRIPTION
**Pull request by ioquatix**
Apparently it's supported now: https://devblogs.microsoft.com/commandline/af_unix-comes-to-windows/

This enables `UNIXSocket`, `UNIXServer` and `UNIXSocket.pair` on Windows.

Because these constants are now available, more tests are run on Windows.

### CODE REVIEW

File: ext/socket/rubysocket.h [hunk 1]
```diff
@@ -67,6 +67,15 @@
 
 #ifdef HAVE_SYS_UN_H
 #  include <sys/un.h>
+#  define HAVE_SOCKADDR_UN 1
```
Comment (nobu) [inline ext/socket/rubysocket.h:70]: `have_type("struct sockaddr_un", "sys/un.h")` defines `HAVE_TYPE_STRUCT_SOCKADDR_UN`.

**Comment by ioquatix** [comment]
While `AF_UNIX` is supported on recent Windows, `socketpair` is not supported. https://github.com/microsoft/WSL/issues/4240

We can try to emulate it.

File: win32/win32.c [hunk 2]
```diff
@@ -4059,8 +4083,7 @@ socketpair_internal(int af, int type, int protocol, SOCKET *sv)
         errno = EPROTOTYPE;
         return -1;
     }
-
-    sv[0] = (SOCKET)INVALID_HANDLE_VALUE;
+OCKET)INVALID_HANDLE_VALUE;
```
Comment (ioquatix) [inline win32/win32.c:4086]:
```suggestion

     sv[0] = (SOCKET)INVALID_HANDLE_VALUE;
```

File: win32/win32.c [hunk 3]
```diff
@@ -4060,7 +4084,7 @@ socketpair_internal(int af, int type, int protocol, SOCKET *sv)
         return -1;
     }
 
-    sv[0] = (SOCKET)INVALID_HANDLE_VALUE;
+     sv[0] = (SOCKET)INVALID_HANDLE_VALUE;
```
Comment (ioquatix) [inline win32/win32.c:4087]:
```suggestion
    sv[0] = (SOCKET)INVALID_HANDLE_VALUE;
```

**Comment by ioquatix** [comment]
@larskanis is there any chance you can help me with this? It would reduce the compatibility differences between windows and linux and implementing it enables a lot more tests. There are some issues with the current PR that I can't figure out because I don't know how to set up a proper Windows development environment. If this is something you can help me with, I've got Windows 11 set up in a VM but can't figure out how to get it all working into a proper build environment.

**Comment by larskanis** [comment]
@ioquatix Thank you for working on this! Did you implement this with just the Github Action runs?

The build environment that I use is just a standard RubyInstaller-Devkit installation and git-for-windows. Ruby core development can be done either in `cmd` like:

```cmd
ridk enable ucrt64
cd c:\work\ruby
sh configure
make
```

or in MSYS2 bash like:
```cmd
ridk enable ucrt64
bash
cd /c/work/ruby
./configure
make
```
Nothing really special - only much slower than on Linux. I don't have any deeper experience with Visual Studio however.

This PR looks like almost finished. Shall I try to fix the remaining issues tomorrow?

**Comment by ioquatix** [comment]
> Did you implement this with just the Github Action runs?

Yeah, haha, but now I've set up virtual box.

> This PR looks like almost finished. Shall I try to fix the remaining issues tomorrow?

If you can fix this PR I will be very very grateful, there are a lot of things I don't understand about the build process but I've tried to get everything together.

I would really love to see this because it would make testing IO on windows much easier.

The only problem, a lot of tests have assumed `UNIXSocket` defined = unix model (e.g. working fork, working nonblock, etc). We just need to be aware of that and fix the tests.

**Comment by larskanis** [comment]
I pushed [the branch](https://github.com/larskanis/ruby/tree/afunix) a bit further, but lost several hours by trying and debugging the abstract namespace in every possible way, just to notice that it's simply not implemented by Microsoft - although explicitly mentioned as supported in the blog post. They will definitely not catch any developers that way. Also the complete lack of documentation is a no-go for closed source software.

I'm on the way to support `File#socket?` and `File#lstat.socket?` now.

**Comment by ioquatix** [comment]
@larskanis if you want, please feel free to commit to this PR. It should work, but on the off chance it doesn't I've sent you an invite to my fork.

**Comment by ioquatix** [comment]
@larskanis how is it going? Would you like to work on this together some time?

**Comment by larskanis** [comment]
@ioquatix I'm still working on this, although I first had to understand the WIndows reparse points better and had to experiment with them. The ruby code to this is overly complex and not very consistent. Currently the tests starve at `TestFiberIO#test_heavy_read` for some reason. I'll continue the work.

**Comment by ioquatix** [comment]
Nice work, let me know if you want to peer program on it sometime!

**Comment by larskanis** [comment]
@ioquatix It took a while, but IMHO https://github.com/ruby/ruby/compare/master...larskanis:ruby:afunix is ready to be reviewed. I didn't change the base commit, but added my commits with appropriate comments. All tests pass now.
I wasn't able to push directly to this PR and the invitation has already expired. Could you please renew it?

**Comment by ioquatix** [comment]
I resent the invite.

**Comment by ioquatix** [comment]
@larskanis feel free to rebase this branch. I've merged your other PR too.

**Comment by larskanis** [comment]
OK, I rebased this branch to 7b1d23fd295fe8275b313f31ea545f7f6b3f2e62.

File: test/fiber/test_io.rb [hunk 4]
```diff
@@ -64,16 +66,11 @@ def test_heavy_read
   end
 
   def test_epipe_on_read
-    omit "UNIXSocket is not defined!" unless defined?(UNIXSocket)
+    omit unless defined?(UNIXSocket)
+    omit "nonblock=true isn't properly supproted on Wondows" if RUBY_PLATFORM=~/mswin|mingw/
```
Comment (ioquatix) [inline test/fiber/test_io.rb:70]:
```suggestion
    omit "nonblock=true isn't properly supported on Windows" if RUBY_PLATFORM=~/mswin|mingw/
```

File: ext/socket/rubysocket.h [hunk 5]
```diff
@@ -69,6 +72,12 @@
 #  include <sys/un.h>
 #endif
 
+#ifdef HAVE_AFUNIX_H
+// The afunix.h header is borked and requires this header to work:
```
Comment (ioquatix) [inline ext/socket/rubysocket.h:76]:
```suggestion
```

**Comment by ioquatix** [comment]
@larskanis thanks for your amazing work, we managed to make IO just a little bit easier on Windows.

**Pull request merged**